### PR TITLE
Add functionality to create heatmaps for each player in ReplayImportService

### PR DESCRIPTION
**Pull request by Adwitiya-Singh**
Supersedes #32

### CODE REVIEW

File: src/Utility/PlayerMovementGrid.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+namespace App\Utility;
+
+class PlayerMovementGrid
+{
+    /** @var array */
+    private $movement;
+    /** @var int Size of the map in the replay file */
+    private $grid_size;
+    /** @var int Size of the heatmap */
+    private $heatmap_size;
```
Comment (allejo) [inline src/Utility/PlayerMovementGrid.php:21]:
```suggestion
    /** @var int[][] */
    private $movement;
    /** @var int Size of the map in the replay file */
    private $worldSize;
    /** @var int Size of the heatmap */
    private $cellSize;
```

camelCase and let's rename `grid_size` to `worldSize` and `heatmap_size` to `cellSize`

**Review by allejo — CHANGES_REQUESTED** [review]
Loving this! Here are some minor code changes

File: src/Service/ReplayImportService.php [hunk 2]
```diff
@@ -180,11 +194,26 @@ class ReplayImportService
     /** @var array<int, string> A map of flag IDs to flag abbreviations */
     private $flagIDs;
 
-    public function __construct(EntityManagerInterface $em, LoggerInterface $logger, MapThumbnailWriterService $thumbnailWriterService)
+    /** @var array<int, PlayerMovementGrid> A map of flag IDs to flag abbreviations */
+    private $currPlayersHeatMap;
+
+    /** @var int The size of the Heatmap */
+    private $heatMapSize;
```
[the remark this answers is not in the document]
Comment (Adwitiya-Singh) [reply src/Service/ReplayImportService.php:201]: Nope. This is the number of the cells in one row of the heatmap.

File: src/Utility/PlayerMovementGrid.php (hunk 1 above)
Comment (Adwitiya-Singh) [reply src/Utility/PlayerMovementGrid.php:21]: renaming `heatmap_size` to `heatmapSize`

File: src/Service/ReplayImportService.php [hunk 3]
```diff
@@ -180,11 +194,26 @@ class ReplayImportService
...
 
-    public function __construct(EntityManagerInterface $em, LoggerInterface $logger, MapThumbnailWriterService $thumbnailWriterService)
+    /** @var array<string, PlayerMovementGrid> A map of callsigns to their respective movement grids */
+    private $currPlayersHeatMap;
+
+    /** @var int The size of the Heatmap */
+    private $heatMapSize;
+
+    /** @var int The size of the World */
+    private $worldSize;
```
Comment (allejo) [inline src/Service/ReplayImportService.php:204]:
```suggestion
    /** @var int The number of cells in a single row of the heatmap */
    private $heatMapSize;

    /** @var int The size of the world in World Units (wu) */
    private $worldSize;
```

Let's change the doc descriptions for these variables

File: src/Utility/PlayerMovementGrid.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+namespace App\Utility;
+
+class PlayerMovementGrid
+{
+    /** @var array */
+    private $movement;
+    /** @var int Size of the map in the replay file */
+    private $worldSize;
+    /** @var int Size of the heatmap */
+    private $heatmapSize;
```
Comment (allejo) [inline src/Utility/PlayerMovementGrid.php:21]:
```suggestion
    /** @var int[][] */
    private $movement;
    /** @var int Size of the map in the replay file in World Units (wu) */
    private $worldSize;
    /** @var int The number of cells in a single row of the heatmaps */
    private $heatmapSize;
```

File: src/Service/HeatMapWriterService.php [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+     * @param string $GradientMid Mid colour for gradient
+     * @param string $GradientEnd End colour for gradient
+     * @return bool
+     */
+    public function writeHeatMap(
+        PlayerHeatMap $heatMap,
+        int $svgSize,
+        string $GradientStart = self::GRADIENT_FIRST,
+        string $GradientMid = self::GRADIENT_SECOND,
+        string $GradientEnd = self::GRADIENT_THIRD
```
Comment (allejo) [inline src/Service/HeatMapWriterService.php:54]: camelCase

File: src/Service/HeatMapWriterService.php [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    ): bool
+    {
+        $heatmap = $heatMap->getHeatmap();
+        $heatmapSize = count($heatmap);
+        $oldRange = $this->maxval($heatmap);
+        $image = new SVG($svgSize, $svgSize);
+        $doc = $image->getDocument();
+
+        $SquareSize = $svgSize / $heatmapSize;
```
Comment (allejo) [inline src/Service/HeatMapWriterService.php:63]: camelCase

**Review by allejo — CHANGES_REQUESTED** [review]

**Review by allejo — APPROVED** [review]
I love this so much! Thank you dude! I'll get this deployed over the next few days! 😄🥳

**Pull request merged**